libraries/AP_Mount/AP_Mount_Gremsy.cpp

```diff
@@ -0,0 +1,318 @@
+                                                   _gimbal_device_attitude_status.angular_velocity_z,
+                                                   _gimbal_device_attitude_status.failure_flags);
+}
+
+// search for gimbal in GCS_MAVLink routing table
+void AP_Mount_Gremsy::find_gimbal()
+{
+    // do not look for gimbal for first 10 seconds so user may see banner
+    if (AP_HAL::millis() < 10000) {
+        return;
+    }
+
+    // return if search time has has passed
+    if (AP_HAL::millis() > AP_MOUNT_GREMSY_SEARCH_MS) {
+        return;
+    }
+
+    // search for a mavlink enabled gimbal
+    if (!_found_gimbal) {
+        mavlink_channel_t chan;
+        uint8_t sysid, compid;
+        if (GCS_MAVLINK::find_by_mavtype(MAV_TYPE_GIMBAL, sysid, compid, chan)) {
+            if (((_instance == 0) && (compid == MAV_COMP_ID_GIMBAL)) ||
+                ((_instance == 1) && (compid == MAV_COMP_ID_GIMBAL2))) {
+                _found_gimbal = true;
+                _sysid = sysid;
+                _compid = compid;
+                _chan = chan;
+            }
+        } else {
+            // have not yet found a gimbal so return
+            return;
+        }
+    }
+
+    // request GIMBAL_DEVICE_INFORMATION
+    if (!_got_device_info) {
+        uint32_t now_ms = AP_HAL::millis();
+        if (now_ms - _last_devinfo_req_ms > 1000) {
+            _last_devinfo_req_ms = now_ms;
+            request_gimbal_device_information(_chan, _sysid, _compid);
+        }
+        return;
+    }
+
+    // start sending autopilot attitude to gimbal
+    if (start_sending_attitude_to_gimbal()) {
+        _initialised = true;
+    }
+}
+
+// handle GIMBAL_DEVICE_INFORMATION message
+void AP_Mount_Gremsy::handle_gimbal_device_information(const mavlink_message_t &msg)
+{
+    // exit immediately if this is not our message
+    if (msg.sysid != _sysid || msg.compid != _compid) {
+        return;
+    }
+
+    mavlink_gimbal_device_information_t info;
+    mavlink_msg_gimbal_device_information_decode(&msg, &info);
+
+    // set parameter defaults from gimbal information
+    _state._roll_angle_min.set_default(degrees(info.roll_min) * 100);
+    _state._roll_angle_max.set_default(degrees(info.roll_max) * 100);
+    _state._tilt_angle_min.set_default(degrees(info.pitch_min) * 100);
+    _state._tilt_angle_max.set_default(degrees(info.pitch_max) * 100);
+    _state._pan_angle_min.set_default(degrees(info.yaw_min) * 100);
+    _state._pan_angle_max.set_default(degrees(info.yaw_max) * 100);
+
+    const uint8_t fw_ver_major = info.firmware_version & 0x000000FF;
+    const uint8_t fw_ver_minor = (info.firmware_version & 0x0000FF00) >> 8;
+    const uint8_t fw_ver_revision = (info.firmware_version & 0x00FF0000) >> 16;
+    const uint8_t fw_ver_build = (info.firmware_version & 0xFF000000) >> 24;
+
+    // display gimbal info to user
+    gcs().send_text(MAV_SEVERITY_CRITICAL, "Mount: %s %s fw:%u.%u.%u.%u",
```

Comment:
why is this critical? how often does this happen?

Reply:
Txs for the review.  reduced to info and it's only written once when the gimbal is first detected.